zkevm-circuits/src/evm_circuit/execution.rs

```diff
@@ -1045,149 +1075,132 @@ impl<F: Field> ExecutionConfig<F> {
-                    }
+                let mut assign_padding_or_step = |cur_tx_call_step: TxCallStep,
+                                                  next_tx_call_step: Option<TxCallStep>,
+                                                  padding_end: Option<usize>|
+                 -> Result<usize, Error> {
+                    let (_tx, call, step) = cur_tx_call_step;
                     let height = step.execution_state().get_step_height();
-
-                    // Assign the step witness
-                    self.assign_exec_step(
-                        &mut region,
-                        offset,
-                        block,
-                        transaction,
-                        call,
-                        step,
-                        height,
-                        next.copied(),
-                        challenges,
-                        assign_pass,
-                    )?;
-
-                    // q_step logic
-                    self.assign_q_step(&mut region, offset, height)?;
-
-                    offset += height;
-                }
-
-                // part2: assign non-last EndBlock steps when padding needed
-                if !no_padding {
-                    if offset >= evm_rows {
-                        log::error!(
-                            "evm circuit offset larger than padding: {} > {}",
+                    let initial_offset = offset; // start with 0
+
+                    // If padding, assign padding range with (dummy_tx, call, step)
+                    // otherwise, assign one row with cur (tx, call, step), with next (tx, call,
+                    // step) to lookahead
+                    if let Some(padding_end) = padding_end {
+                        // padding_end is the absolute position over all rows,
+                        // must be greater then current offset
+                        if offset >= padding_end {
+                            log::error!(
+                                "evm circuit offset larger than padding: {} > {}",
+                                offset,
+                                padding_end
+                            );
+                            return Err(Error::Synthesis);
+                        }
+                        log::trace!("assign Padding in range [{},{})", offset, padding_end);
+                        self.assign_same_exec_step_in_range(
+                            &mut region,
                             offset,
-                            evm_rows
-                        );
-                        return Err(Error::Synthesis);
+                            padding_end,
+                            block,
+                            chunk,
+                            (&dummy_tx, call, step),
+                            height,
+                            challenges,
+                            assign_pass,
+                        )?;
+                        let padding_start = offset;
```

Comment:
chores: if move `let padding_start = offset;` to line 1123, then 1126, 1127 we can use  `padding_start` and `padding_end` which improved readibility